File: leetcode.com/0011_container_with_most_water/solution.c
#include <stdlib.h>
#include <stdio.h>
#include <stdbool.h>

#define MIN(X, Y) (((X) < (Y)) ? (X) : (Y))

int maxArea(int* height, int heightSize)
{
    int result = 0;
    int begin = 0;
    int end = heightSize - 1;

    while (begin < end) {
        int square = (end - begin) * MIN(height[begin], height[end]);

        if (square > result)
            result = square;

        if (height[begin] < height[end]) {
            begin++;
        } else {
            end--;
        }
    }

    return result;
}

int main()
{
    int size = 15000;
    int* height = malloc(sizeof(int) * size);

    for (int i = 0 ; i < size ; i++)
        height[i] = rand() % 1000;

    printf("%d == 14743480\n", maxArea(height, size));
    return 0;
}

File: leetcode.com/0011_container_with_most_water/solution.go
import "math"

func maxArea(height []int) int {
    max := 0
    left := 0
    right := len(height) - 1
    
    for left < right {
        square := (right - left) * int(math.Min(float64(height[left]), float64(height[right])))
        
        if square > max {
            max = square
        }
        
        if (height[left] > height[right]) {
            right--
        } else {
            left++
        }
    }
    
    return max
}


